### DISCUSSION TITLE
Shelly Dimmer in KNX integrieren

### DISCUSSION THREAD

**Discussion by brutella** [brutella/hkknx-public#108] [Show and tell]
Ein [Shelly Dimmer](https://shelly.cloud) wird verwendet um Lampen zu schalten und dimmen. Der Dimmer ist klein und kann in einer Unterputzdose untergebracht werden. Die Steuerung erfolgt entweder über die Shelly App, die Weboberfläche oder über die HTTP API.

Im folgenden Skript wird die HTTP API verwendet um zwei Shelly Dimmers (`shellydimmer-f39d06` und `shellydimmer-d0a6da`) in KNX zu integrieren. Für jeden Dimmer werden 4 Gruppenadressen angelegt:

- shellydimmer-f39d06
    - 1/0/1 Schalten 
    - 1/0/2 Schalten Status
    - 1/0/3 Dimmen 
    - 1/0/4 Dimmen Status 
- shellydimmer-d0a6da
    - 1/1/1 Schalten 
    - 1/1/2 Schalten Status
    - 1/1/3 Dimmen 
    - 1/1/4 Dimmen Status

Dann wird für den jeden Shelly über die Weboberfläche unter **Actions** die URLs *OUTPUT SWITCHED ON URL* und *OUTPUT SWITCHED OFF URL* eingetragen. Die URL besteht aus der Adresse von hkknx und dem Namen des Gerätes.

Am Beispiel für den Shelly `shellydimmer-f39d06` sind die URLs wie folgt.
- OUTPUT SWITCHED ON URL: http://hkknx.local:8090/shellydimmer-f39d06/on
- OUTPUT SWITCHED OFF URL: http://hkknx.local:8090/shellydimmer-f39d06/off

```go
// This scripts makes Shelly light dimmers available to KNX.
// 
// The built-in [HTTP API][1] in Shelly dimmers is used to change the 
// state and brightness. 
// [1]: https://shelly-api-docs.shelly.cloud/#shelly-bulb-light-0
// 
// On the other hand a Shelly dimmers sends an HTTP request to the 
// script when it is turned on. To make this work you have to define
// an "OUTPUT SWITCHED ON URL" and "OUTPUT SWITCHED OFF URL" in the 
// Shelly web UI under "Actions" .
// 
// OUTPUT SWITCHED ON URL  → http://{ip-address-of-hkknx}:8090/{shelly-dimmer-name}/on
// OUTPUT SWITCHED OFF URL → http://{ip-address-of-hkknx}:8090/{shelly-dimmer-name}/off
// Ex. http://hkknx.local:8090/shellydimmer-f39d06/off
var hkknx = import("hkknx")
var time = import("time")
var http = import("net/http")
var fmt = import("fmt")
var ioutil = import("io/ioutil")
var json = import("encoding/json")

// shellies contains a list of group address.
// 
// The order is as follows:
// [0] state
// [1] state status
// [2] brightness
// [3] brightness status
shellies = {
	"shellydimmer-f39d06": ["1/0/1", "1/0/2", "1/0/3", "1/0/4"],
	"shellydimmer-d0a6da": ["1/1/1", "1/1/2", "1/1/3", "1/1/4"],
}

mux = http.NewServeMux()
observeShelly = func(name, switchAddr, switchStateAddr, brightnessAddr, brightnessStateAddr) {
	mux.HandleFunc(fmt.Sprintf("/%s/on", name), func(w, r) {
		println(name, "turned on")
		hkknx.GroupWrite(true, switchStateAddr)
		w.WriteHeader(204)
	})
	
	mux.HandleFunc(fmt.Sprintf("/%s/off", name), func(w, r) {
		println(name, "turned off")
		hkknx.GroupWrite(false, switchStateAddr)
		w.WriteHeader(204)
	})
	sw = hkknx.GroupWriteBoolNotify(switchAddr)
	brightness = hkknx.GroupWriteNotify(brightnessAddr)
	swState = hkknx.GroupReadNotify(switchStateAddr)
	brightnessState = hkknx.GroupReadNotify(switchStateAddr)
	hostname = fmt.Sprintf("%s.local", name)
	
	getStatus = func() {
		// Fetch the state via HTTP GET request
		url = fmt.Sprintf("http://%s/light/0", hostname)
		resp, err = http.DefaultClient.Get(url)
		if err != nil {
			return nil, err
		}
		
		// Read the response body
		buf, err = ioutil.ReadAll(resp.Body)
		if err != nil {
			return nil, err
		}
		
		// Decode the json content
		data = make(map[string]interface)
		err = json.Unmarshal(buf, &data)
		return data, err
	}
	
	go func() {
		for {
			select {
			case on = <-sw: // Wait for write telegrams
				cmd = "off"
				if on { cmd = "on" }
				url = fmt.Sprintf("http://%s/light/0?turn=%s", hostname, cmd)
				println(url)
				resp, err = http.DefaultClient.Get(url)
				if err != nil {
					println(err)
				}
				
			case buf = <-brightness: // Wait for write telegrams
				value = hkknx.ParseDPT5001(buf)
				url = fmt.Sprintf("http://%s/light/0?brightness=%.0f", hostname, value)
				println(url)
				resp, err = http.DefaultClient.Get(url)
				if err != nil {
					println(err)
				}
				
			case <-swState: // Wait for read telegrams
				data, err = getStatus()
				if err != nil {
					println(err)
					continue
				}
				
				// Respond to read telegram
				hkknx.GroupResponse(data["ison"], stateAddr)
			
			case <-brightnessState: // Wait for read telegrams
				data, err = getStatus()
				if err != nil {
					println(err)
					continue
				}
				
				// Respond to read telegram
				hkknx.GroupResponse(hkknx.DPT5001(data["brightness"]), brightnessStateAddr)
			}
		}	
	}()	
}

for name, addrs in shellies {
	observeShelly(name, addrs[0], addrs[1], addrs[2], addrs[3])
}

ListenAndServe(":8090", mux)
```

Als weiteren Schritt können man mit den KNX Gruppenadressen ein HomeKit Gerät in hkknx anlegen. Somit wäre der Shelly dann auch über HomeKit steuerbar.

**Comment by thorsten-gehrig** [comment]
Hi
könnte man über den selben Weg iko´s vom EDOMI über die passende Web-Schnittstelle integrieren?
(lesend und schreibend?)
Gruß
Thorsten

**Comment by brutella** [comment]
Welche Schnittstelle(n) stellt EDOMI dafür zur Verfügung?

**Reply by thorsten-gehrig** [reply]
Hi
zum einen MQTT (mit passenden Plugins) - oder nativ per HTTP.
Hier ein Auszug aus der Hilfe für den Bereich "Remotezugriff:"
---
Der Zugriff auf ein per Fernzugriff freigegebenes KO erfolgt über folgende URL:

Abfragen eines KO-Wertes:
http://<IP-Adresse>/remote/?login=<Login>&pass=<Passwort>&koid=<KO-ID>

Setzen eines KO-Wertes:
http://<IP-Adresse>/remote/?login=<Login>&pass=<Passwort>&koid=<KO-ID>&kovalue=<KO-Wert>

<IP-Adresse>: IP-Adresse des EDOMI-Servers
<Login>: Login (s.u.)
<Passwort>: Passwort (s.u.)

<KO-ID>: ID des Kommunikationsobjekts (nicht die Gruppenadresse!)

<KO-Wert>: Wert, auf den das KO ggf. gesetzt werden soll
wird der Parameter "&kovalue=<KO-Wert>" nicht angegben, wird der aktuelle KO-Wert von EDOMI zurückgegeben
Hinweis: Bei einer KNX-GA erfolgt kein(!) Read-Request, es wird stets der letzte bekannte GA-Wert aus dem internen Prozessabbild zurückgegeben.

Beispiele:

http://192.168.0.10/remote/?login=remote&pass=remote&koid=3
fragt das System-KO [3] (Server-IP) ab, der Rückgabewert wäre z.B. "1.2.3.4" (ohne "")

http://192.168.0.10/remote/?login=remote&pass=remote&koid=100&kovalue=EDOMI
setzt das KO [100] auf den Wert "EDOMI", der Rückgabewert wäre z.B. "OK;100;EDOMI;" (ohne "")


Rückgabewerte

Je nach Parametern und Status wird der o.g. HTTP-Request wie folgt beantwortet:

<KO-Wert>
das abgefragte KO hat aktuell den Wert <KO-Wert>

OK;<KO-ID>;<KO-Wert>;
das KO <KO-ID> wird/wurde auf den Wert <KO-Wert> gesetzt

ERROR;<KO-ID>;FORBIDDEN;
das KO <KO-ID> ist nicht für den Fernzugriff freigegeben oder existiert nicht

ERROR;LOGIN;
die Zugangsdaten sind nicht korrekt
---
Ich hoffe das Hilft :-)
Gruß
Thorsten

**Reply by brutella** [reply]
Handelt es sich bei den HTTP Requests um GET oder POST Requests?
